File: clients/android/arizona/src/main/kotlin/dev/arizona/client/Wire.kt
package dev.arizona.client

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Pure wire-protocol logic, with no Android/Compose dependency, so it is
 * unit-testable on the JVM. Mirrors the browser worker's `zipTemplate`/
 * `resolveHtml` (assets/js/arizona-core.js) and the reference client
 * (e2e/utils/native_client.js): a fingerprint cache plus an interleaver that
 * stitches statics and dynamics into a JSON string.
 */

/** Diff op codes (mirror src/arizona.hrl). */
object Op {
    const val TEXT = 0
    const val SET_ATTR = 1
    const val REM_ATTR = 2
    const val REMOVE_NODE = 4
    const val INSERT = 5
    const val REMOVE = 6
    const val ITEM_PATCH = 7
    const val REPLACE = 8
    const val MOVE = 9
}

/** Effect command op codes (mirror include/arizona_effect.hrl). */
object Effect {
    const val PUSH_EVENT = 0
    const val NAVIGATE = 10
}

/**
 * Fingerprint-cache bound (mirrors `FP_CACHE_MAX` in assets/js/arizona-core.js).
 * A fingerprint hashes a template's statics, so every deploy mints a new key and
 * orphans the old one: unbounded, a long-lived client accumulates one generation
 * per deploy. Evicting is never wrong -- the cache is content-addressed, so the
 * server re-sends the statics for any key the client did not announce, and a miss
 * costs bytes only -- but it is only SAFE between connections: once a socket has
 * announced a key the server stops shipping that template's statics, so dropping
 * it mid-connection would leave a payload the client cannot resolve. The prune
 * therefore runs at announce time; the map is free to grow within a session.
 */
const val FP_CACHE_MAX = 1000

/**
 * Caches a fingerprint's statics so later frames can omit them (the server
 * sends `{f, d}` once `{f, s, d}` has been seen, deduplicated by fingerprint).
 */
class FingerprintCache {
    private data class Entry(val statics: List<String>, val t: Int?)

    // Access-ordered, so iteration is least-recently-used FIRST and the
    // announce-time prune just drops from the front.
    private val byFp = LinkedHashMap<String, Entry>(16, 0.75f, true)

    /** Statics for [payload], caching them if present, else from a prior frame. */
    fun statics(payload: JsonObject): List<String> {
        val f = payload.getValue("f").jsonPrimitive.content
        payload["s"]?.let { s ->
            byFp[f] = Entry(
                statics = s.jsonArray.map { it.jsonPrimitive.content },
                t = payload["t"]?.jsonPrimitive?.int,
            )
        }
        return byFp[f]?.statics ?: error("uncached fingerprint: $f")
    }

    /**
     * The cached fingerprints, most-recently-used LAST, after pruning down to
     * [limit]. Announced on every socket open so the server can omit statics this
     * client already holds.
     */
    fun announce(limit: Int = FP_CACHE_MAX): List<String> {
        val oldest = byFp.keys.iterator()
        while (byFp.size > limit) {
            oldest.next()
            oldest.remove()
        }
        return byFp.keys.toList()
    }
}

/**
 * Stitches a `{f, s, d}` payload into a JSON string, JSON-encoding each dynamic
 * value (string -> quoted, number -> as-is) -- the native counterpart of the
 * browser's string-concatenating zip. Recurses into nested `{f, s, d}` payloads
 * (child components) and each `{t: 0, ...}` item-lists.
 */
class Interleaver(private val cache: FingerprintCache) {

    fun interleave(payload: JsonObject): String = interleaveWith(cache.statics(payload), payload.getValue("d").jsonArray)

    /**
     * Decode an op payload to its JSON form: a `{t:0}` each-list -> a JSON array,
     * a `{f, s, d}` template -> a JSON object, a scalar -> itself. Used to
     * materialize OP_INSERT items and OP_TEXT content.
     */
    fun decode(payload: JsonElement): String = encodeValue(payload)

    private fun interleaveWith(statics: List<String>, dynamics: JsonArray): String {
        val sb = StringBuilder(statics[0])
        dynamics.forEachIndexed { i, v ->
            sb.append(encodeValue(v)).append(statics[i + 1])
        }
        return sb.toString()
    }

    private fun encodeValue(v: JsonElement): String =
        when {
            // An ?each expansion -> a JSON array of items the parent splices in.
            v is JsonObject && v["t"]?.jsonPrimitive?.intOrNull == 0 -> {
                val s = cache.statics(v)
                val items = v.getValue("d").jsonArray.map { interleaveWith(s, it.jsonArray) }
                items.joinToString(prefix = "[", postfix = "]", separator = ",")
            }
            // A nested {f, s, d} template (child component).
            v is JsonObject -> interleave(v)
            // A scalar: JsonPrimitive.toString() is its JSON form (quoted/bare).
            else -> (v as JsonPrimitive).toString()
        }
}

File: clients/ios/Sources/AzWire/Wire.swift
import Foundation

/// Diff op codes (mirror src/arizona.hrl).
public enum Op {
    public static let text = 0
    public static let setAttr = 1
    public static let remAttr = 2
    public static let update = 3
    public static let removeNode = 4
    public static let insert = 5
    public static let remove = 6
    public static let itemPatch = 7
    public static let replace = 8
    public static let move = 9
}

/// Effect command op codes (mirror include/arizona_effect.hrl). Only the two
/// portable effects are meaningful to a native client; the rest are web-only.
public enum Effect {
    public static let pushEvent = 0
    public static let navigate = 10
}

/// A wire-shape failure the op applier recovers from by skipping ONE op. Swift's
/// counterpart of the exceptions the Kotlin and JS clients catch per op: a
/// `preconditionFailure` here would take the whole app down over a single slot.
public enum WireError: Error {
    case malformed(String)
}

/// Fingerprint-cache bound (mirrors `FP_CACHE_MAX` in assets/js/arizona-core.js).
/// A fingerprint hashes a template's statics, so every deploy mints a new key and
/// orphans the old one: unbounded, a long-lived client accumulates one generation
/// per deploy. Evicting is never wrong -- the cache is content-addressed, so the
/// server re-sends the statics for any key the client did not announce, and a miss
/// costs bytes only -- but it is only SAFE between connections: once a socket has
/// announced a key the server stops shipping that template's statics, so dropping
/// it mid-connection would leave a payload the client cannot resolve. The prune
/// therefore runs at announce time; the map is free to grow within a session.
public let fpCacheMax = 1000

/// Caches a fingerprint's statics so later frames can omit them (the server
/// sends `{f, d}` once `{f, s, d}` has been seen, deduplicated by fingerprint).
public final class FingerprintCache {
    private struct Entry {
        let statics: [String]
        let t: Int?
        // Monotonic use stamp: ranks the announce-time prune (Swift dictionaries
        // are unordered, so the stamp is what Android gets from an access-ordered
        // LinkedHashMap and JS from Map insertion order).
        var used: UInt64
    }

    private var byFp: [String: Entry] = [:]
    private var clock: UInt64 = 0

    public init() {}

    /// Statics for `payload`, caching them if present, else from a prior frame.
    public func statics(_ payload: [String: JSONValue]) throws -> [String] {
        guard let f = payload["f"]?.stringValue else {
            throw WireError.malformed("payload has no fingerprint: \(payload)")
        }
        clock += 1
        if let s = payload["s"]?.arrayValue {
            byFp[f] = Entry(
                statics: s.map { $0.stringValue ?? "" }, t: payload["t"]?.intValue, used: clock)
        } else {
            byFp[f]?.used = clock
        }
        guard let entry = byFp[f] else { throw WireError.malformed("uncached fingerprint: \(f)") }
        return entry.statics
    }

    /// The cached fingerprints, most-recently-used LAST, after pruning down to
    /// `limit`. Announced on every socket open so the server can omit statics this
    /// client already holds.
    public func announce(limit: Int = fpCacheMax) -> [String] {
        let ranked = byFp.sorted { $0.value.used < $1.value.used }.map { $0.key }
        guard ranked.count > limit else { return ranked }
        for key in ranked.prefix(ranked.count - limit) { byFp[key] = nil }
        return Array(ranked.suffix(limit))
    }
}

/// Stitches a `{f, s, d}` payload into a JSON value, JSON-encoding each dynamic
/// (string -> quoted, number -> as-is) -- the native counterpart of the browser's
/// string-concatenating zip. Recurses into nested `{f, s, d}` payloads (child
/// components) and `{t: 0, ...}` item-lists. Mirrors Android's `Interleaver`.
public final class Interleaver {
    private let cache: FingerprintCache

    public init(_ cache: FingerprintCache) {
        self.cache = cache
    }

    /// Interleave a `{f, s, d}` payload into its rendered JSON value (an object).
    ///
    /// Throws (rather than trapping) on any payload it cannot resolve -- an
    /// uncached fingerprint, a statics/dynamics length mismatch, a non-object --
    /// so the op applier can skip that one op instead of the app dying. See
    /// `WireError`.
    public func interleave(_ payload: JSONValue) throws -> JSONValue {
        try JSONValue.parseChecked(interleaveString(payload))
    }

    /// Decode an op payload to its JSON value: a `{t:0}` each-list -> a JSON
    /// array, a `{f, s, d}` template -> an object, a scalar -> itself. Used to
    /// materialize OP_INSERT items and OP_TEXT/OP_UPDATE content. (Returns a
    /// value directly rather than Android's JSON string, sidestepping top-level
    /// scalar-fragment parsing; behavior is identical.)
    public func decode(_ payload: JSONValue) throws -> JSONValue {
        if case .object = payload { return try JSONValue.parseChecked(encodeValue(payload)) }
        return payload
    }

    private func interleaveString(_ payload: JSONValue) throws -> String {
        guard case let .object(obj) = payload else {
            throw WireError.malformed("interleave expects an object: \(payload)")
        }
        return try interleaveWith(cache.statics(obj), payload["d"]?.arrayValue ?? [])
    }

    private func interleaveWith(_ statics: [String], _ dynamics: [JSONValue]) throws -> String {
        guard statics.count == dynamics.count + 1 else {
            throw WireError.malformed("\(statics.count) statics for \(dynamics.count) dynamics")
        }
        var out = statics[0]
        for (i, v) in dynamics.enumerated() {
            out += try encodeValue(v)
            out += statics[i + 1]
        }
        return out
    }

    private func encodeValue(_ v: JSONValue) throws -> String {
        guard case let .object(obj) = v else { return v.serialized } // a scalar
        if obj["t"]?.intValue == 0 {
            // An ?each expansion -> a JSON array of items the parent splices in.
            let s = try cache.statics(obj)
            var items: [String] = []
            for itemD in obj["d"]?.arrayValue ?? [] {
                items.append(try interleaveWith(s, itemD.arrayValue ?? []))
            }
            return "[" + items.joined(separator: ",") + "]"
        }
        return try interleaveString(v) // a nested {f, s, d} template
    }
}
